File: frontend/src/lib/types/responses.ts
// API response types matching FastAPI schemas

export type GenerationStatus =
  | "parsing"
  | "transforming"
  | "generating_images"
  | "generating_output"
  | "uploading"
  | "complete"
  | "cache_hit"
  | "error";

export interface ProgressEvent {
  status: GenerationStatus;
  progress: number;
  message?: string;
}

export interface CompletionMetadata {
  title: string;
  pages: number;
  slides: number;
  images_generated: number;
}

export interface CompleteEvent {
  status: "complete";
  progress: 100;
  download_url: string;
  file_path: string;
  expires_in: number;
  metadata: CompletionMetadata;
}

export interface CacheHitEvent {
  status: "cache_hit";
  progress: 100;
  download_url: string;
  file_path: string;
  expires_in: number;
  cached_at: string;
}

export interface ErrorEvent {
  status: "error";
  error: string;
  code: string;
}

export type GenerationEvent =
  | ProgressEvent
  | CompleteEvent
  | CacheHitEvent
  | ErrorEvent;

export interface UploadResponse {
  file_id: string;
  filename: string;
  size: number;
  mime_type: string;
  expires_in: number;
}

export interface HealthResponse {
  status: string;
  version: string;
}

export function isCompleteEvent(event: GenerationEvent): event is CompleteEvent {
  return event.status === "complete";
}

export function isCacheHitEvent(event: GenerationEvent): event is CacheHitEvent {
  return event.status === "cache_hit";
}

export function isErrorEvent(event: GenerationEvent): event is ErrorEvent {
  return event.status === "error";
}

export function isProgressEvent(event: GenerationEvent): event is ProgressEvent {
  return !isCompleteEvent(event) && !isCacheHitEvent(event) && !isErrorEvent(event);
}

File: backend/doc_generator/infrastructure/api/schemas/responses.py
"""Pydantic response models for the API."""

from enum import Enum

from pydantic import BaseModel


class GenerationStatus(str, Enum):
    PARSING = "parsing"
    TRANSFORMING = "transforming"
    GENERATING_IMAGES = "generating_images"
    GENERATING_OUTPUT = "generating_output"
    UPLOADING = "uploading"
    COMPLETE = "complete"
    CACHE_HIT = "cache_hit"
    ERROR = "error"


class ProgressEvent(BaseModel):
    status: GenerationStatus
    progress: int
    message: str = ""


class CompletionMetadata(BaseModel):
    title: str
    pages: int = 0
    slides: int = 0
    images_generated: int = 0


class CompleteEvent(BaseModel):
    status: GenerationStatus = GenerationStatus.COMPLETE
    progress: int = 100
    download_url: str
    file_path: str
    expires_in: int = 3600
    metadata: CompletionMetadata


class CacheHitEvent(BaseModel):
    status: GenerationStatus = GenerationStatus.CACHE_HIT
    progress: int = 100
    download_url: str
    file_path: str
    expires_in: int = 3600
    cached_at: str


class ErrorEvent(BaseModel):
    status: GenerationStatus = GenerationStatus.ERROR
    error: str
    code: str


class UploadResponse(BaseModel):
    file_id: str
    filename: str
    size: int
    mime_type: str
    expires_in: int = 3600


class HealthResponse(BaseModel):
    status: str = "healthy"
    version: str
